File: epic_ops/epic_ops/ccl.py
import torch


@torch.no_grad()
def connected_components_labeling(
    indices: torch.Tensor,
    edges: torch.Tensor,
    compacted: bool = True,
) -> torch.Tensor:
    indices = indices.contiguous()
    edges = edges.contiguous()

    return torch.ops.epic_ops.connected_components_labeling(
        indices, edges, compacted
    )

File: epic_ops/src/epic_ops/ccl.cpp
#include <torch/library.h>
#include "epic_ops/ccl.h"

namespace epic_ops::ccl {

at::Tensor connected_components_labeling(
    const at::Tensor& indices,
    const at::Tensor& edges,
    bool compacted) {
  static auto op = c10::Dispatcher::singleton()
                       .findSchemaOrThrow("epic_ops::connected_components_labeling", "")
                       .typed<decltype(connected_components_labeling)>();
  return op.call(indices, edges, compacted);
}

TORCH_LIBRARY_FRAGMENT(epic_ops, m) {
  m.def(TORCH_SELECTIVE_SCHEMA(
      "epic_ops::connected_components_labeling(Tensor indices, Tensor edges, "
      "bool compacted) -> Tensor"));
}

} // namespace epic_ops::ccl
